Aggregate diffs

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/21035
## Description

Changes diff view for descriptions etc. to show the diff of the corresponding aggregated journal.
Thereby clicking on a diff link in the activity tab (e.g. of the details pane) leads to a correct diff for the displayed activity (in case the description was changed twice).

This PR also unifies the methods `details`, `get_changes` and `changed_data` of the `Journal` (which are all aliases of eachother). That means that occurences of `changed_data` have been replaced by `details`. In the long term the goal is to remove one or two methods of this alias chain.
## Note

Since the legacy WP view does not aggregate journals, it will now link to unmatching diffs. This is deliberate, since the legacy WP view is going to vanish in OP 5.0.

Commit:
replace changed_data by details

working towards making the alias obsolete...

NobodysNightmare committed:
display diff of containing aggregated journal

... instead of displaying the diff for the requested journal only

- this should provide the best forward compatibility of links
- even when aggregation timeframes are changed intermittently

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-aggregated-diffs.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-aggregated-diffs.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

ulferts commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -41,6 +41,12 @@
+    # Returns the aggregated journal that contains the specified (vanilla) journal.
+    def for_journal(vanilla_journal)
+      Journal::AggregatedJournal.aggregated_journals(journable: vanilla_journal.journable)
+                  .detect { |journal| journal.version >= vanilla_journal.version }
```
By relying on 'detect', you are fetching all the journals from the DB, instantiate them in ruby, and then filter for only one of them.
You could instead use a where clause to filter in the db.

NobodysNightmare replied:
true.
and now that my other PR is merged, there is already a helper method for that SQL.... will do that!

oliverguenther commented on app/models/journal_manager.rb:
```diff
@@ -136,7 +136,7 @@ def self.recreate_initial_journal(type, journal, changed_data)
       journal.data = create_journal_data journal.id, type, changed_data.except(:id)
     else
-      journal.changed_data = changed_data
+      journal.details = changed_data
```
Not sure If you already had a look at the specs and the reason is immediately clear - I had to look why this line fails in all related specs, and it is due to https://github.com/NobodysNightmare/openproject/blob/feature/aggregated_diffs/app/models/journal.rb#L59-L66 not being changed correctly.

NobodysNightmare committed:
filter for right journal on SQL level

oliverguenther commented:
Works as expected. LGTM and waiting on travis. :+1:

oliverguenther commented:
Hmm.. I've checked the failing `work_package_notifications_spec` and the deliveries are sent correctly locally and no specs fail.

As jobs are not delayed in the test environment, I suspected that this has something to do with the notifications, however they are blocking as well, right? :confused:

NobodysNightmare committed:
ensure aggregation is disabled when testing diff

alternative would be to ensure that the change is done "late" enough or by a different user.
I hope this way is the most explicit one.

NobodysNightmare commented:
Seems like the test I modified is now a little flickering...

This is probably due to MySQL only storing the creation date with second precision.
I discussed with @ulferts that it should suffice to increase the precision of the affected column to subseconds (~milli/micro) to circumvent the problem.
This is also neccessary to effectively allow disabling aggregation.

Since I don't want to introduce a flickerer, I will do the apropriate fixes in this PR.

oliverguenther commented:
One could add a second setting to disable the aggregation entirely, since you do not know the exact time the delayed_job will run, even when `run_at` is now. Also not sure what the difference is for `0.from_now` to the time provided by [`db_time_now`](http://www.rubydoc.info/gems/delayed_job/2.1.4/Delayed/Backend/ActiveRecord/Job#db_time_now-class_method) used by DJ as default

NobodysNightmare commented:
The issues here are quite unrelated to delayed job, aren't they?

My problem is that MySQL will store `created_at` with an insufficent precision. When I then later test whether the time difference between two journals is `> 0`, I get a wrong result, because of the low precision. I see no DJ involvement here.

oliverguenther commented:
Ah, I was following the wrong premise on this test then. But If you change the precision on the `created_table`, wouldn't you also have to update `sql_beyond_aggregation_time?` with a higher precision for the aggregation time diff, e.g., with `MICROSECOND`?

NobodysNightmare commented:
@oliverguenther Yep, the update of the corresponding SQL was part of the issue. Thanks for pointing that out, I may have missed it for a longer time otherwise :)

NobodysNightmare committed:
enforce disabled aggregations

due to limited time resolution we can't assume that a setting of 0 will implicitly disable aggregation
